File: 1407-group-the-people-given-the-group-size-they-belong-to/1407-group-the-people-given-the-group-size-they-belong-to.py
class Solution:
    def groupThePeople(self, groupSizes: List[int]) -> List[List[int]]:
        d=defaultdict(list)
        res=[]
        for i,x in enumerate(groupSizes) :
            d[x].append(i)
            if len(d[x])==x :
                res.append(d[x])
                d.pop(x)
        return res
             

File: 1407-group-the-people-given-the-group-size-they-belong-to/1407-group-the-people-given-the-group-size-they-belong-to.java
class Solution {
    public List<List<Integer>> groupThePeople(int[] groupSizes) {
        Map<Integer, List<Integer>> sizeToGroup = new HashMap<>();
        List<List<Integer>> res = new ArrayList<>();
        
        for (int i = 0; i < groupSizes.length; i++) 
        {
            int size = groupSizes[i];
            sizeToGroup.putIfAbsent(size, new ArrayList<>());
            List<Integer> group = sizeToGroup.get(size);
            group.add(i);
            
            // map.computeIfAbsent(groupSizes[i], k -> new ArrayList()).add(i);

            if (group.size() == size) 
            {
                res.add(group);
                sizeToGroup.put(size, new ArrayList<>());

                // sizeToGroup.remove(size);
            }
        }
        
        return res;
    }
}